Add support and sample backend for dictionary-based logging

dcpleung opened the pull request:
Dictionary-based logging, instead of human readable texts, outputs the log messages in binary format. This binary format encodes arguments to formatted strings in their native storage formats which can be more compact than their text equivalents. For statically defined strings (including the format strings and any string arguments), references to the ELF file are encoded
instead of the whole strings. A dictionary created at build time contains the mappings between these references and the actual strings. This allows the offline parser to obtain the strings from the dictionary to parse the log messages. This binary format allows a more compact representation of log messages in certain scenarios. However, this requires the use of an offline parser and is not as intuitive to use as text-based log messages.

dcpleung committed:
nios2: add _image_rodata_start/end to linker script

This adds the _image_rodata_start and _image_rodata_end
symbols to NIOS2's linker script.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
doc: add documentation for dictionary-based logging

This adds new documentation for dictionary-based logging on
usage and low-level information.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
sparc: add _image_rodata_start/end to linker script

This adds the _image_rodata_start and _image_rodata_end
symbols to SPARC's arch linker script.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung commented:
CI's failure to build documentation is not related to this PR.

nordic-krch commented:
@dcpleung please look at #30353. I started to work on logging subsystem overhaul to enable missing features which cannot be achieved with current design. The main change will be that log message will be self contained (string duplication will be inside the message). Messages will no longer be fragmented. This will significantly simplify dictionary based logging because dictionary backend will just transmit raw message as is.

To make that happen there is also an ongoing work in cbprintf to allow that (#30675) by introducing a concept of string package. A string package contains serialized arguments. Strings are treated specially, they can be stored by address or by value. There is 1 one header which indicates mode and address or content follows. Cbprintf will support formatting a string from a package in parallel to standard formatting from va_list.

nashif commented:
> @dcpleung please look at #30353. I started to work on logging subsystem overhaul to enable missing features which cannot be achieved with current design. The main change will be that log message will be self contained (string duplication will be inside the message). Messages will no longer be fragmented. This will significantly simplify dictionary based logging because dictionary backend will just transmit raw message as is.

@nordic-krch what are you suggesting here? Wait for the overhaul? What is the timeline?

nordic-krch commented:
@nashif i've opened draft PR #31535

dcpleung committed:
logging: simplify is_rodata() for NIOS2, RISC-V and SPARC

NIOS2, RISC-V and SPARC are using _image_rodata_start/_end in
their linker scripts to mark the boundaries of rodata. So
they no loner need special treatment.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
lib/os: cbprintf_packaged: simplify rodata check

NIOS2 is using _image_rodata_start/_end in its linker script
to mark the boundaries of rodata. So they no loner need
special treatment.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
logging: add support for dictionary based logging

This adds dictionary based logging support. Dictionary based
logging is binary based where one big difference is that
static strings are stored as pointers instead of the whole
string. This results in reduced space requirements for
storing log messages in certain scenairos.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
logging: add UART backend for dictionary based logging

This adds a new UART backend for dictionary based logging,
where this can output binary data in both binary and
hexidecimal strings.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

pabigot commented:
Capturing a comment from the API telecon:  The stored representation must be able to intersperse log records with updates to the dictionary that was used to produce the log record content, so that stores of log data that span multiple sessions are decipherable when the log spans firmware updates that change the dictionary.

In short: archived logs must be self-describing.  To accomplish this there must be a way to fetch and emit the current dictionary whenever records begin to be written to an output (back-end).

dcpleung commented:
This is rebased on top of #31535.

jukkar commented on doc/guides/logging/dictionary.rst:
```diff
@@ -0,0 +1,208 @@
+
+Here is an example of how to use dictionary-based logging.
+
+1. :file:`samples/hello_world/src/main.c` can be modified for demostration
+   purpose.
```
Instead of instructions to modify an existing sample, why not create a new sample in `samples/subsys/logging` and have all the necessary code there.

dcpleung replied:
Having a new sample app is better... what was I thinking...? :)

jukkar commented on doc/guides/logging/dictionary.rst:
```diff
@@ -0,0 +1,208 @@
+      SeaBIOS (version zephyr-v1.0.0-0-g31d4e0e-dirty-20200714_234759-fv-az50-zephyr)
+      Booting from ROM..000004000000000000a9d3000004000000e05810001e6010009f5f10000000030000000000001309010003000000c4511000d45e1000000802000000000000fc2f010002000000b45110000020030000000000003c51010003000000a05110005c5b10000018020000000000001378010002000000945110000020050000000000005099010005000000785110005c5b1000010000000200000000200500000000000069cb0100050000005c5110005c5b100010000000110000000 ...
+
+5. Copy the hexidecimal string into a temporary file ``/tmp/serial.log``:
```
Copying the log to temp file sounds quite cumbersome. Could the output be piped directly to log_parser.py instead?

dcpleung replied:
Piping the output from west build or west run sends all data to the parser. There is no easy way to distinguish between normal text and log data from a stream of text. Though output QEMU, the log would be captured through serial or flash storage which would only contain log binary data. So I am not sure if we want to make the log parser more complicated so it could parse the build output.

jukkar replied:
Sure, I understand. It is just that copying data from console to temp file is not very practical in real life. Could the log parser just monitor some start bytes/marker and ignore all the data it receives before that. That sounds like an easy solution to this issue, what do you think?

dcpleung replied:
Added a separator so the parser can find where the log begins.

dcpleung committed:
samples: logging: add a sample app for dictionary-based logging

This adds a sample application on dictionary-based logging.
The README file includes instruction on how to run the log parser
to generate human readable log messages.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

andyross reviewed (approved):
Finally getting to look at this seriously.  Looks clean, though I'll admit that I mostly just skimmed the python.

One thing I think we should add is a version cookie somewhere, even just one encoded as a noop message at the beginning of the stream or something.  It's a 100% certainty we're going to want to come back and change this protocol and having parsers in the wild be able to discriminate output from deployed device with different firmware versions is going to be important.

Also, and this really isn't in scope, but this is being hurt very badly by our architectural decision to make "output hardware method" and "output processing layer" into the same "backend" abstraction.   Now we have to have TWO backends for every one of our platforms, and do the same engineering repeatedly to unify the implementations to avoid cutting and pasting.  Really we should have a distinction between "transport" and "protocol" plugins...

nordic-krch commented on subsys/logging/dictionary/log_dict_backend_uart.c:
```diff
@@ -0,0 +1,101 @@
+	uart_dev = device_get_binding(CONFIG_UART_CONSOLE_ON_DEV_NAME);
+	__ASSERT_NO_MSG((void *)uart_dev);
+
+#if defined(CONFIG_LOG_DICTIONARY_BACKEND_UART_HEX)
```
using `IS_ENABLED()` would remove ifdef from here and variable declaration.

nordic-krch commented on include/logging/log_output_dict.h:
```diff
@@ -0,0 +1,75 @@
+ * Output header for one dictionary based log message.
+ */
+struct log_dict_output_normal_msg_hdr_t {
+	uint8_t type;
```
have you considered reusing log_msg2 directly. There are 2 bits at the start which are used by mpsc_pbuf. They cannot be modified in the provided buffer since they are still in use but if first word (descriptor) would be copied then they could be reused to indicate message type.

This is rather an optimization proposal not necessary for the first iteration.

dcpleung replied:
At first I was planning to use `struct log_msg2` directly, but decided not to. This is simply to ease future maintenance. Any changes (even trivial) to `struct log_msg2` would need to be reflected on the Python parser. Also, any alignment and/or padding done to the struct also needs to be done on the Python side which would create a bit more work during arch bring-up. Possible outcome would be that the Python parser is not updated and dictionary-based logging would then be broken on the new arch. (though, it might be just me being paranoid.)

nordic-krch reviewed (commented):
Looks good, few minor comments

nordic-krch commented on samples/subsys/logging/dictionary/README.rst:
```diff
@@ -0,0 +1,73 @@
+
+   .. code-block:: none
+
+      [     54185] <none> hello_world: *** Booting Zephyr OS build zephyr-v2.5.0-2075-g1e5790192a3c  ***
```
printk are now without prefix. README should be updated.

dcpleung committed:
lib: os: cbprintf: fix different signedness comparison

One of the loop is using int as loop variable but it's being
compared against size_t. Some compiler configuration has
-Wsign-compare turned on by default which will complain about
"comparison of integer expressions of different signedness".
So fix it by changing the loop variable to be size_t also.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

tejlmand commented on subsys/logging/dictionary/Kconfig:
```diff
@@ -0,0 +1,17 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+
+config LOG_DICTIONARY_SUPPORT
```
We already have a `Logging --> Output Formatting` menu here:
https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/logging/Kconfig.formatting

So please move this config in there, for example after the MIPI_SYST formatter.
https://github.com/zephyrproject-rtos/zephyr/blob/2cff32b6c4c68a9ed24375ee42b1d3aa79b7195d/subsys/logging/Kconfig.formatting#L24-L29

tejlmand commented on subsys/logging/dictionary/Kconfig:
```diff
@@ -15,3 +15,30 @@ config LOG_DICTIONARY_SUPPORT
 	  Normally this is selected by the backend automatically and does
 	  not need to be enabled manually.
+
+config LOG_DICTIONARY_BACKEND_UART
+	bool "Enable UART backend for dictionary based logging"
+	depends on LOG2 && UART_CONSOLE
+	select LOG_DICTIONARY_SUPPORT
+	help
+	  When enabled backend is using UART to output dictionary based logs.
```
We also have backends here:
https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/logging/Kconfig.backends

Actually, when I build `samples/subsys/logging/dictionary` for nrf52840dk_nrf52840 board, I get both this dictionary backend uart enabled but also RTT, so where will my logs go ?
```
$ cmake -GNinja -DBOARD=nrf52840dk_nrf52840  -Bbuild_52 samples/subsys/logging/dictionary/
$ grep 'CONFIG_LOG_.*BACKEND' build_52/zephyr/.config
# CONFIG_LOG_BACKEND_UART is not set
CONFIG_LOG_BACKEND_RTT=y
...
CONFIG_LOG_DICTIONARY_BACKEND_UART=y
```

And why is a dedicated `LOG_DICTIONARY_BACKEND_UART` needed, how does it differ from, for example using `LOG_BACKEND_UART=y` with dictionary formatting ?

dcpleung replied:
I have rolled the dictionary based UART output into the upper level UART backend, where kconfig is used to choose the output.

As for which backend, the log messages will go to all enabled backends, which is by design.

tejlmand reviewed (changes requested):
General impression is good, but the Kconfig needs some cleanup.

dcpleung commented on samples/subsys/logging/dictionary/src/main.c:
```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright (c) 2012-2014 Wind River Systems, Inc.
```
The original is from `hello_world` so here is to preserve copyright.

dcpleung commented:
Blocked by #34558

tejlmand commented on subsys/logging/Kconfig.backends:
```diff
@@ -3,20 +3,56 @@
+	  Output in text.
+
 config LOG_BACKEND_UART_SYST_ENABLE
 	bool "Enable UART syst backend"
```
Thanks for making a choice out of this.
But when making a choice, I think it would be nice if this prompt was updated so the look-n-feel in whole choice group is aligned.

```suggestion
	bool "MIPI SyS-T"
```

tejlmand commented on subsys/logging/Kconfig.backends:
```diff
@@ -3,20 +3,56 @@
 menu "Backends"
 
-config LOG_BACKEND_UART
+menuconfig LOG_BACKEND_UART
```
Using a menuconfig config here  will give you an extra level when having to change backend.

First entering the `menu` (that only has a single choice), and thereafter entering the choice selection itself which is an additional layer.

The choice itself creates a menu for selection, and is only visible when `LOG_BACKEND_UART=y`.

```suggestion
config LOG_BACKEND_UART
```

tejlmand reviewed (changes requested):
Looks very good. Thanks for the cleanup into a choice.

Two minor nits.

tejlmand reviewed (approved):
LGTM

marc-hb commented:
Dictionary for printk() in the Linux kernel:
https://lwn.net/Articles/857148/
https://lwn.net/ml/linux-kernel/cover.1621338324.git.chris@chrisdown.name/
